llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,270 @@
+
+  return Reg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  bool Modified = false;
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  BitVector UsedPhyRegs(TRI->getNumRegs());
+  SmallVector<std::pair<MCPhysReg, MCPhysReg>, 8> ReplacementRegs;
+
+  BitVector ForbiddenPhysRegs = getDefinedPhysRegs(MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+
+      MCPhysReg PreSwapPhyReg = VRM->getPhys(Reg);
+      if (isWorthRenaming(Reg, MI, UsedPhyRegs) &&
+          replaceReg(Reg, ForbiddenPhysRegs)) {
+
+        LLVM_DEBUG(dbgs() << MI);
+        Modified = true;
+        ReplacementRegs.push_back(
+            std::make_pair(PreSwapPhyReg, VRM->getPhys(Reg)));
+
+      } else {
+        // keep track of already visted physical Regs
+        UsedPhyRegs[VRM->getPhys(Reg)] = true;
+      }
+    }
+  }
+  // dump Reassignment
+  LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+             dbgs() << "\n"; for (std::pair<MCPhysReg, MCPhysReg> &RegPair
+                                  : ReplacementRegs) {
+               dbgs() << printReg(RegPair.first, TRI, 0, MRI) << " -> "
+                      << printReg(RegPair.second, TRI, 0, MRI) << "\n";
+             } dbgs() << "\n");
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWorthRenaming(const Register &Reg,
+                                        const MachineInstr &MI,
+                                        const BitVector &UsedPhyRegs) const {
+  assert(Reg.isVirtual() && MRI->getRegClass(Reg));
+  if (!UsedPhyRegs[VRM->getPhys(Reg)])
+    return false;
+  return true;
+}
+
+BitVector AIEWawRegRewriter::getDefinedPhysRegs(const MachineBasicBlock *MBB) {
+  BitVector BlockedPhysRegs(TRI->getNumRegs());
+
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      addAliasRegs(BlockedPhysRegs, getAssignedPhysReg(Op));
```

Comment:
getAssignedPhysReg can return NoRegister. Is that ok for addAliasReg?

Reply:
already handeled in a fixup